modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestBodyStream.java

```diff
@@ -33,25 +34,29 @@
 public class Netty4HttpRequestBodyStream implements HttpBody.Stream {
 
     private final Channel channel;
-    private final ChannelFutureListener closeListener = future -> doClose();
     private final List<ChunkHandler> tracingHandlers = new ArrayList<>(4);
     private final ThreadContext threadContext;
+    private final AutoReadSync.Handle autoRead;
+    private final ThreadWatchdog.ActivityTracker activityTracker;
     private ByteBuf buf;
     private boolean requested = false;
-    private boolean closing = false;
+    private final ChannelFutureListener closeListener = future -> doClose();
     private HttpBody.ChunkHandler handler;
     private ThreadContext.StoredContext requestContext;
+    private boolean hasLast = false;
+    private boolean closed = false;
 
     // used in tests
     private volatile int bufSize = 0;
-    private volatile boolean hasLast = false;
 
-    public Netty4HttpRequestBodyStream(Channel channel, ThreadContext threadContext) {
+    public Netty4HttpRequestBodyStream(Channel channel, ThreadContext threadContext, ThreadWatchdog.ActivityTracker activityTracker) {
         this.channel = channel;
         this.threadContext = threadContext;
+        this.activityTracker = activityTracker;
         this.requestContext = threadContext.newStoredContext();
+        this.autoRead = AutoReadSync.getHandle(channel);
```

Comment:
I'm uncomfortable with dynamically acquiring a new auto-read flag for each body stream in the channel. We're only reading one of these at once, can we do something more obviously bounded here?

Reply:
We can read fairly large chunk from network that can contain multiple HTTP requests. We will dispatch each request headers to `RestHandlers` while holding content for each of them in the `Netty4HttpRequestBodyStream` buffer. So it's possible that multiple stream handlers can exists in the channel that toggle off `auto-read` at the same time.

The ultimate and clean solution that I see here is moving `Netty4HttpHeaderValidator` to the rest layer. Place `FlowControlHandler` in front of `HttpContentDecompressor` that will hold all incoming http parts, and toggle auto-read only in the `Netty4HttpRequestBodyStream`. The bitset toggle I introduced here only for the time being until we move auth stuff.